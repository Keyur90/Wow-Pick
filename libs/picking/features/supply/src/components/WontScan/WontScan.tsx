import {
  Button,
  Dialog,
  DialogActions,
  DialogContent,
  FormControl,
  makeStyles,
  MenuItem,
  Select,
} from '@material-ui/core';
import { ToteItem } from '@rf2/picking/data/api-contracts';
import { convertKgToGrams, isWeightedQuantityArticle, useInputQuantityValidation } from '@rf2/shared/utility';
import { QuantityField } from '@rf2/ui';
import Big from 'big.js';
import React, { useEffect, useState } from 'react';

interface WontScanPropTypes {
  onConfirm: (quantityValue: number, isSubbed: boolean) => void;
  onClose: () => void;
  toteItem: ToteItem;
  qtyFromBarcodeScan: number;
  isPrimary?: boolean;
}

const useDialogContentStyles = makeStyles((theme) => ({
  root: {
    padding: theme.spacing(0, 2.5),
    '&:first-child': {
      paddingTop: theme.spacing(3.5),
    },
  },
}));

const useDialogActionStyles = makeStyles((theme) => ({
  root: {
    padding: theme.spacing(3, 2.5),
  },
}));

const useStyles = makeStyles((theme) => ({
  formControl: {
    marginTop: theme.spacing(2),
    width: '100%',
  },
}));

const WontScan: React.FC<WontScanPropTypes> = ({
  onConfirm,
  onClose,
  toteItem,
  qtyFromBarcodeScan,
  isPrimary = true,
}) => {
  const dialogContentClasses = useDialogContentStyles();
  const dialogActionClasses = useDialogActionStyles();
  const classes = useStyles();
  const { orderedQuantity, totalSuppliedQuantity } = toteItem;
  const [hasSubmitted, setHasSubmitted] = useState(false);
  const isWeighted = isWeightedQuantityArticle(toteItem);
  const tmpOrderedQuantity = new Big(orderedQuantity);
  let defaultValue = qtyFromBarcodeScan;
  if (defaultValue == null) {
    defaultValue = tmpOrderedQuantity.minus(totalSuppliedQuantity).gt(0)
      ? tmpOrderedQuantity.minus(totalSuppliedQuantity).toNumber()
      : 0;
  }
  const [inputValue, setInputValue] = useState(isWeighted ? convertKgToGrams(defaultValue) : defaultValue);
  const { inputError } = useInputQuantityValidation(inputValue, toteItem);

  const [selectedArticleId, setSelectedArticleId] = useState(
    isPrimary ? toteItem.articleId : toteItem.secondaryArticleId
  );
  const [isSubbed, setIsSubbed] = useState(false);

  useEffect(() => {
    if (selectedArticleId === toteItem.articleId) {
      setIsSubbed(false);
    } else {
      setIsSubbed(true);
    }
  }, [selectedArticleId]);

  const handleInputFocus = () => {
    setHasSubmitted(false);
  };

  const handleInputBlur = (evt) => {
    setInputValue(evt?.target?.value);
  };

  const handleConfirm = () => {
    setHasSubmitted(true);
    if (!inputError) {
      onConfirm(inputValue, isSubbed);
    }
  };

  const articleList = [
    { id: toteItem.articleId, type: 'primary', display: `${toteItem.articleId} (Ordered item)` },
    { id: toteItem.secondaryArticleId, type: 'silent', display: `${toteItem.secondaryArticleId} (Equivalent item)` },
  ];

  const onArticleSelected = (event: React.ChangeEvent<{ value: unknown }>) => {
    setSelectedArticleId(articleList.find(({ id }) => event.target.value === id)?.id);
  };

  return (
    <Dialog open={true} fullWidth maxWidth="sm">
      <DialogContent classes={dialogContentClasses}>
        <QuantityField
          name="suppliedQuantity"
          label={isWeighted ? 'Enter the weight in grams' : 'Enter the supplied quantity'}
          defaultValue={!isWeighted ? inputValue : ''}
          onFocus={handleInputFocus}
          onBlur={handleInputBlur}
          error={hasSubmitted && !!inputError}
          errorText={inputError}
          startAdornment={isWeighted ? 'g' : null}
        />
        {toteItem.secondaryArticleId ? (
          <FormControl className={classes.formControl}>
            <Select value={selectedArticleId} onChange={onArticleSelected} defaultChecked={true}>
              {articleList.map(({ id, display }) => (
                <MenuItem key={id} value={id}>
                  {display}
                </MenuItem>
              ))}
            </Select>
          </FormControl>
        ) : null}
      </DialogContent>
      <DialogActions classes={dialogActionClasses}>
        <Button color="secondary" onClick={onClose}>
          CANCEL
        </Button>
        <Button variant="contained" color="secondary" onClick={handleConfirm}>
          CONFIRM
        </Button>
      </DialogActions>
    </Dialog>
  );
};

export { WontScan };
