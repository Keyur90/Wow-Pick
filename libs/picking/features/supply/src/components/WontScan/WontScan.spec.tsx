import { MockedProvider } from '@apollo/client/testing';
import { render } from '@testing-library/react';
import React from 'react';
import { WontScan } from './WontScan';

test('renders WontScan', () => {
  const toteItem = {
    id: '28391',
    toteId: '4',
    itemIndex: 0,
    orderNo: '140031681',
    articleId: '28391',
    aisle: '9',
    bay: '56',
    shelf: '2',
    orderedQuantity: 3,
    currentPage: 5,
    stockOnHand: 23,
    totalPages: 20,
    version: '0.0.1',
    price: 4.0,
    secondaryArticleId: '124577',
    personalShopperNotes:
      "Please only make a substitution if you really have too. As these are the best pumpkin cubes I've ever had.",
    suppliedDetails: [],
    totalSuppliedQuantity: 0,
    article: {
      id: '123123',
      pricingUnit: 'EACH',
      brand: 'Brand',
      description: 'Description',
    },
  };
  const { baseElement } = render(
    <MockedProvider mocks={null} addTypename={false}>
      <WontScan onConfirm={() => null} onClose={() => null} toteItem={toteItem} />
    </MockedProvider>
  );

  expect(baseElement).toMatchSnapshot();
});
