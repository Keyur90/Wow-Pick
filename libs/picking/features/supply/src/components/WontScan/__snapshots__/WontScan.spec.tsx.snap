// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders WontScan 1`] = `
<body
  style="padding-right: 0px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  />
  <div
    class="MuiDialog-root"
    role="presentation"
    style="position: fixed; z-index: 1300; right: 0px; bottom: 0px; top: 0px; left: 0px;"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-test="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiDialog-container MuiDialog-scrollPaper"
      role="none presentation"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      tabindex="-1"
    >
      <div
        class="MuiPaper-root MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthSm MuiDialog-paperFullWidth MuiPaper-elevation24 MuiPaper-rounded"
        role="dialog"
      >
        <div
          class="MuiDialogContent-root makeStyles-root-1"
        >
          <div
            class="MuiFormControl-root MuiTextField-root makeStyles-root-4 MuiFormControl-fullWidth"
          >
            <label
              class="MuiFormLabel-root MuiFormLabel-colorSecondary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-filled Mui-focused Mui-focused"
              data-shrink="true"
              for="suppliedQuantity"
              id="suppliedQuantity-label"
            >
              Enter the supplied quantity
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorSecondary MuiOutlinedInput-colorSecondary MuiInputBase-fullWidth Mui-focused Mui-focused MuiInputBase-formControl"
            >
              <input
                aria-invalid="false"
                autocomplete="off"
                class="MuiInputBase-input MuiOutlinedInput-input"
                id="suppliedQuantity"
                inputmode="numeric"
                name="suppliedQuantity"
                pattern="[0-9]*"
                type="text"
                value="3"
              />
              <fieldset
                aria-hidden="true"
                class="PrivateNotchedOutline-root-5 MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="PrivateNotchedOutline-legendLabelled-7 PrivateNotchedOutline-legendNotched-8"
                >
                  <span>
                    Enter the supplied quantity
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <div
            class="MuiFormControl-root makeStyles-formControl-3"
          >
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
            >
              <div
                aria-haspopup="listbox"
                class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
                role="button"
                tabindex="0"
              >
                28391 (Ordered item)
              </div>
              <input
                aria-hidden="true"
                class="MuiSelect-nativeInput"
                tabindex="-1"
                value="28391"
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSelect-icon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
            </div>
          </div>
        </div>
        <div
          class="MuiDialogActions-root makeStyles-root-2 MuiDialogActions-spacing"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textSecondary"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              CANCEL
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              CONFIRM
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
    <div
      data-test="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
